core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java

```diff
@@ -160,7 +202,22 @@ public AllocationService(Settings settings, AllocationDeciders allocationDecider
         if (!reroute(allocation)) {
             return new RoutingAllocation.Result(false, clusterState.routingTable());
         }
-        return new RoutingAllocation.Result(true, new RoutingTable.Builder().updateNodes(routingNodes).build().validateRaiseException(clusterState.metaData()));
+        RoutingTable routingTable = new RoutingTable.Builder().updateNodes(routingNodes).build().validateRaiseException(clusterState.metaData());
+        RoutingAllocation.Result result = new RoutingAllocation.Result(true, routingTable);
+        logClusterHealthStateChange(
+                new ClusterStateHealth(clusterState),
+                new ClusterStateHealth(clusterState.getMetaData(), routingTable),
+                reason
+        );
+        return result;
+    }
+
+    private void logClusterHealthStateChange(ClusterStateHealth previousStateHealth, ClusterStateHealth newStateHealth, String reason) {
```

Comment:
I think we can make this method even friendlier with the following signature: `logClusterHealthStateChange(ClusterStateHealth previousStateHealth, Result result, String reason)` .